SYRCoSE article (#74)
Event-Structures/event-struct: papers/SYRCoSE/regmachine-eventstruct.tex

we define the causality relation of the register machine's event structure
as a transitive closure of two relations~---~\emph{program order} 
and \emph{reads-from}, denoted as $\lPO$ and $\lRF$ correspondingly. 

$$ \caOrd \ \defeq \ (\lPO \cup \lRF)^* $$

Program order tracks precedence of events within a single thread. 
Reads-from relation captures the flow of values from 
write events to read events, and ensures that values 
do not appear out of thin-air~\cite{Boehm-Demsky:MSPC14, Lahav-al:PLDI17}.  

In order to construct $\lPO$ and $\lRF$ incrementally
we represent them via their inverse covering functions $\lfPO$ and $\lfRF$. 

\begin{definition}[Covering]
  Let $\leqslant$ be a partial order. 
  Then $\lessdot$ is covering relation \wrt $\leqslant$ whenever 
  $x \lessdot y$ is true if and only if $x < y$ and 
  there is no $z$ \sth $x < z$ and $z < y$.
  A (non-deterministic) function $f : A \fun \pwset{A}$ is a covering function if
  its corresponding relation, \ie ${\frel{f} \defeq \set{\tup{x, y} ~|~ y \in f(x)}}$, 
  is a covering relation.
\end{definition}
 
We use the inverse covering function because it is 

> more convinient in our setting. Indeed, the semanics adds new event

Reviewer: ```suggestion
more convenient in our setting. Indeed, the semantics adds a new event
```

> at each step. Then it is convinient to require that in addition 

Reviewer: ```suggestion
at each step. Then it is convenient to require that in addition 
```

write event is added to the structure. 
After that the load instruction from the middle thread is executed. 
Since there are two matching write events in the event structure, 
two conflicting reads are conjoined to the event structure. 
Note that the events can be added non-deterministically in any order
respecting the causality. We could have first execute the rightmost thread
and add write $\wlab{x}{2}$ before $\wlab{x}{1}$, 
or add the read with label $\rlab{x}{2}$ 
before another read $\rlab{x}{1}$.  

\input{fig/eventstruct-sem.tex}

The rules of operational semantics constructing 
the event structure are presented on~\cref{fig:eventstruct-sem}.
The first auxiliary rule \ESAddEventRule adds a new event, sets its 
label, $\lPO$ and $\lRF$ predecessors. 
The \ESIdleRule handles the case when a thread of 
the register machine performs an internal step 
without any side effect. 
It chooses an event $e$ together with 
the thread state $s$ corresponding to it
and performs one step reduction to a new state $s'$.
It then updates the mapping of events to thread states.   
The last two rules \ESStoreRule and \ESLoadRule

> correspond to store and load perfromed by some thread.  

Reviewer: ```suggestion
correspond to store and load performed by some thread.  
```
